### PR TITLE
Feature & Refactor: write XC matrix in KS-wfc representation and unify matrix-writing functions

### LINKED ISSUE #3443: output XC matrix
Output the XC matrix in Kohn-Sham orbital representation $V^\text{xc}_{ij}=\braket{\psi_i|\hat{V}^\text{xc}|\psi_j}$ , useful for GW calculation.

### PR DESCRIPTION
**Pull request by maki49**
fix #3443
### Feature
The XC matrix in KS-wfc representation with the size of (nbands*nbands) will be useful for the subsequent GW calculation.
Set `out_mat_xc` to 1 in INPUT file to output this in `OUT.$suffix/k-$ik-Vxc`.

### Refactor
Unify the several functions in `write_HS.h` into one. Now we can use `ModuleIO::save_mat` to write a 2d-block parallelized square matrix:
- with any dimension
- whether double or complex<double>
- whether triangle or not

### CODE REVIEW

File: source/module_io/write_Vxc.hpp [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        // double total_energy = 0.0;
+        // double exx_energy = 0.0;
+        // ======test=======
+        for (int ik = 0;ik < kv.nks;++ik)
+        {
+            ModuleBase::GlobalFunc::ZEROS(vxc_k_ao.data(), pv->nloc);
+            int is = GlobalV::CURRENT_SPIN = kv.isk[ik];
+            dynamic_cast<hamilt::OperatorLCAO<TK, TR>*>(vxcs_op_ao[is])->contributeHk(ik);
+#ifdef __EXX
+            if (GlobalC::exx_info.info_global.cal_exx) vexx_op_ao.contributeHk(ik);
```
Comment (dyzheng) [inline source/module_io/write_Vxc.hpp:180]: EXX term is also writed, I suggest you describe this in annotation of this function and documentation.

Comment (dyzheng) [reply source/module_io/write_Vxc.hpp:180]: Another question is why don't you consider strong correlation term DFTU?

File: source/module_io/write_Vxc.hpp [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        ModuleBase::TITLE("ModuleIO", "write_Vxc");
+        const Parallel_Orbitals* pv = lm.ParaV;
+        int nbands = wg.nc;
+        // 1. real-space xc potential
+        // ModuleBase::matrix vr_xc(nspin, chg.nrxx);
+        double etxc = 0.0;
+        double vtxc = 0.0;
+        // elecstate::PotXC* potxc(&rho_basis, &etxc, vtxc, nullptr);
+        // potxc.cal_v_eff(&chg, &ucell, vr_xc);
+        elecstate::Potential* potxc = new elecstate::Potential(&rhod_basis, &rho_basis, &ucell, &vloc, &sf, &etxc, &vtxc);
```
Comment (dyzheng) [inline source/module_io/write_Vxc.hpp:133]: Why don't you try to get elecstate::PotXC from existed elecstate::Potential, which has member vector<PotBase*>.
Then the function "cal_v_eff" can help you.

Comment (dyzheng) [reply source/module_io/write_Vxc.hpp:133]: Sorry, I found Veff<OperatorLCAO> rely on Potential , maybe if this operator can only rely on PotBase will make the code more flexible, I will consider this question.

Comment (maki49) [reply source/module_io/write_Vxc.hpp:133]: > Why don't you try to get elecstate::PotXC from existed elecstate::Potential, which has member vector<PotBase*>. Then the function "cal_v_eff" can help you.

Yes. `Veff::contributeHR()` depends on `Potential::get_effective_v` which is not implemented in `PotBase`, so I have to use `Potential` if I don't want to copy the code in `Veff::contributeHR()`.

File: source/module_io/write_Vxc.hpp (hunk 1 above)
Comment (maki49) [reply source/module_io/write_Vxc.hpp:180]: > Another question is why don't you consider strong correlation term DFTU?

I will talk to @minyez about this. The problem is whether to output the well-defined XC part or the part that GW needs: If DFTU term is included, should DeePKS term also be included?

Comment (maki49) [reply source/module_io/write_Vxc.hpp:180]: We decided to include the DFTU term but not include the DeePKS term (as noted in the document).

**Review by dyzheng — APPROVED** [review]
It is a good feature.

**Pull request merged**